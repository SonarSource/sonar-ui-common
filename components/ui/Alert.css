/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.alert {
  margin-bottom: var(--gridSize);
  border: 1px solid;
  border-radius: 2px;
}

.alert.is-inline {
  display: inline-flex;
}

.alert:empty {
  display: none;
}

.alert-inner {
  display: flex;
  align-items: stretch;
}

.alert-inner.is-banner {
  min-width: var(--minPageWidth);
  max-width: var(--maxPageWidth);
  margin-left: auto;
  margin-right: auto;
  padding-left: var(--pagePadding);
  padding-right: var(--pagePadding);
  box-sizing: border-box;
}

.alert-icon {
  flex: 0 0 auto;
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(4 * var(--gridSize));
  border-right: 1px solid;
}

.alert-icon.is-banner {
  width: calc(2 * var(--gridSize));
  border-right: none;
}

.alert-content {
  flex: 1 0 0;
  padding: var(--gridSize) calc(2 * var(--gridSize));
  overflow: auto;
  text-align: left;
}

.alert-content a,
.alert-content .button-link {
  border-color: var(--darkBlue);
}

.alert-error {
  border-color: var(--alertBorderError);
  background-color: var(--alertBackgroundError);
  color: var(--alertTextError);
}

.alert-error .alert-icon {
  border-color: var(--alertBorderError);
}

.alert-warning {
  border-color: var(--alertBorderWarning);
  background-color: var(--alertBackgroundWarning);
  color: var(--alertTextWarning);
}

.alert-warning .alert-icon {
  border-color: var(--alertBorderWarning);
}

.alert-info {
  border-color: var(--alertBorderInfo);
  background-color: var(--alertBackgroundInfo);
  color: var(--alertTextInfo);
}

.alert-info .alert-icon {
  border-color: var(--alertBorderInfo);
}

.alert-success {
  border-color: var(--alertBorderSuccess);
  background-color: var(--alertBackgroundSuccess);
  color: var(--alertTextSuccess);
}

.alert-success .alert-icon {
  border-color: var(--alertBorderSuccess);
}
