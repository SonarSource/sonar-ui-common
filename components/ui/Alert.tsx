/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import { translate } from '../../helpers/l10n';
import { Dict } from '../../types/types';
import AlertErrorIcon from '../icons/AlertErrorIcon';
import AlertSuccessIcon from '../icons/AlertSuccessIcon';
import AlertWarnIcon from '../icons/AlertWarnIcon';
import InfoIcon from '../icons/InfoIcon';
import { css, styled, Theme, ThemedProps, useTheme } from '../theme';
import DeferredSpinner from './DeferredSpinner';

type AlertDisplay = 'banner' | 'inline' | 'block';
type AlertVariant = 'error' | 'warning' | 'success' | 'info' | 'loading';

export interface AlertProps {
  display?: AlertDisplay;
  variant: AlertVariant;
}

interface AlertVariantInformation {
  icon: JSX.Element;
  color: string;
  borderColor: string;
  backGroundColor: string;
}

const StyledAlertIcon = styled.div<{ isBanner: boolean; variantInfo: AlertVariantInformation }>`
  flex: 0 0 auto;
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(${(props) => (props.isBanner ? 2 : 4)} * ${(props) => props.theme.sizes.gridSize});
  border-right: ${(props) => (!props.isBanner ? '1px solid' : 'none')};
  border-color: ${(props) => props.variantInfo.borderColor};
`;

const StyledAlertContent = styled.div`
  flex: 1 1 auto;
  overflow: auto;
  text-align: left;
  padding: ${(props) => props.theme.sizes.gridSize}
    calc(2 * ${(props) => props.theme.sizes.gridSize});
`;

const alertInnerIsBannerMixin = ({ theme }: ThemedProps) => css`
  min-width: ${theme.sizes.minPageWidth};
  max-width: ${theme.sizes.maxPageWidth};
  margin-left: auto;
  margin-right: auto;
  padding-left: ${theme.sizes.pagePadding};
  padding-right: ${theme.sizes.pagePadding};
  box-sizing: border-box;
`;

const StyledAlertInner = styled.div<{ isBanner: boolean }>`
  display: flex;
  align-items: stretch;
  ${(props) => (props.isBanner ? alertInnerIsBannerMixin : null)}
`;

const StyledAlert = styled.div<{ isInline: boolean; variantInfo: AlertVariantInformation }>`
  border: 1px solid;
  border-radius: 2px;
  margin-bottom: ${(props) => props.theme.sizes.gridSize};
  border-color: ${(props) => props.variantInfo.borderColor};
  background-color: ${(props) => props.variantInfo.backGroundColor};
  color: ${(props) => props.variantInfo.color};
  display: ${(props) => (props.isInline ? 'inline-block' : 'block')};

  :empty {
    display: none;
  }

  a,
  .button-link {
    border-color: ${(props) => props.theme.colors.darkBlue};
  }
`;

function getAlertVariantInfo({ colors }: Theme, variant: AlertVariant): AlertVariantInformation {
  const variantList: Dict<AlertVariantInformation> = {
    error: {
      icon: <AlertErrorIcon fill={colors.alertIconError} />,
      color: colors.alertTextError,
      borderColor: colors.alertBorderError,
      backGroundColor: colors.alertBackgroundError,
    },
    warning: {
      icon: <AlertWarnIcon fill={colors.alertIconWarning} />,
      color: colors.alertTextWarning,
      borderColor: colors.alertBorderWarning,
      backGroundColor: colors.alertBackgroundWarning,
    },
    success: {
      icon: <AlertSuccessIcon fill={colors.alertIconSuccess} />,
      color: colors.alertTextSuccess,
      borderColor: colors.alertBorderSuccess,
      backGroundColor: colors.alertBackgroundSuccess,
    },
    info: {
      icon: <InfoIcon fill={colors.alertIconInfo} />,
      color: colors.alertTextInfo,
      borderColor: colors.alertBorderInfo,
      backGroundColor: colors.alertBackgroundInfo,
    },
    loading: {
      icon: <DeferredSpinner timeout={0} />,
      color: colors.alertTextInfo,
      borderColor: colors.alertBorderInfo,
      backGroundColor: colors.alertBackgroundInfo,
    },
  };

  return variantList[variant];
}

export function Alert(props: AlertProps & React.HTMLAttributes<HTMLDivElement>) {
  const theme = useTheme();
  const { className, display, variant, ...domProps } = props;
  const isInline = display === 'inline';
  const isBanner = display === 'banner';
  const variantInfo = getAlertVariantInfo(theme, variant);

  return (
    <StyledAlert
      className={classNames('alert', className)}
      isInline={isInline}
      role="alert"
      variantInfo={variantInfo}
      {...domProps}>
      <StyledAlertInner isBanner={isBanner}>
        <StyledAlertIcon
          aria-label={translate('alert.tooltip', variant)}
          isBanner={isBanner}
          variantInfo={variantInfo}>
          {variantInfo.icon}
        </StyledAlertIcon>
        <StyledAlertContent className="alert-content">{props.children}</StyledAlertContent>
      </StyledAlertInner>
    </StyledAlert>
  );
}
