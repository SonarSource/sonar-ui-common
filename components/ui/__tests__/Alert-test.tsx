/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { Alert, AlertProps } from '../Alert';

it('should render properly', () => {
  expect(shallowRender({ variant: 'error' })).toMatchSnapshot();
});

it('verification of all variants of alert', () => {
  const variants: AlertProps['variant'][] = ['error', 'warning', 'success', 'info', 'loading'];
  variants.forEach((variant) => {
    const wrapper = shallowRender({ variant });
    expect(wrapper.prop('variantInfo')).toMatchSnapshot();
  });
});

it('should render inline alert', () => {
  expect(shallowRender({ display: 'inline' }).find('Styled(div)[isInline=true]').exists()).toBe(
    true
  );
});

it('should render banner alert', () => {
  expect(shallowRender({ display: 'banner' }).find('Styled(div)[isBanner=true]').exists()).toBe(
    true
  );
});

it('should render banner alert with correct css', () => {
  expect(shallowRender({ display: 'banner' }).render()).toMatchSnapshot();
});

function shallowRender(props: Partial<AlertProps>) {
  return shallow(
    <Alert className="alert-test" id="error-message" variant="error" {...props}>
      This is an error!
    </Alert>
  );
}
