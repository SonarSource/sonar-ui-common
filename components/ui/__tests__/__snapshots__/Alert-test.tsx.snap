// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render banner alert with correct css 1`] = `
.emotion-3 {
  border: 1px solid;
  border-radius: 2px;
  margin-bottom: 8px;
  border-color: #ebccd1;
  background-color: #f2dede;
  color: #862422;
  display: block;
}

.emotion-3:empty {
  display: none;
}

.emotion-3 a,
.emotion-3 .button-link {
  border-color: #236a97;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  min-width: 1080px;
  max-width: 1320px;
  margin-left: auto;
  margin-right: auto;
  padding-left: 20px;
  padding-right: 20px;
  box-sizing: border-box;
}

.emotion-0 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: calc(2 * 8px);
  border-right: none;
  border-color: #ebccd1;
}

.emotion-1 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  overflow: auto;
  text-align: left;
  padding: 8px calc(2 * 8px);
}

<div
  class="alert alert-test emotion-3"
  id="error-message"
  role="alert"
>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-0"
    >
      <svg
        height="16"
        space="preserve"
        style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:1.41421"
        version="1.1"
        viewBox="0 0 16 16"
        width="16"
        xlink="http://www.w3.org/1999/xlink"
      >
        <path
          d="M11.402 10.018q0-0.232-0.17-0.402l-1.616-1.616 1.616-1.616q0.17-0.17 0.17-0.402 0-0.241-0.17-0.411l-0.804-0.804q-0.17-0.17-0.411-0.17-0.232 0-0.402 0.17l-1.616 1.616-1.616-1.616q-0.17-0.17-0.402-0.17-0.241 0-0.411 0.17l-0.804 0.804q-0.17 0.17-0.17 0.411 0 0.232 0.17 0.402l1.616 1.616-1.616 1.616q-0.17 0.17-0.17 0.402 0 0.241 0.17 0.411l0.804 0.804q0.17 0.17 0.411 0.17 0.232 0 0.402-0.17l1.616-1.616 1.616 1.616q0.17 0.17 0.402 0.17 0.241 0 0.411-0.17l0.804-0.804q0.17-0.17 0.17-0.411zM14.857 8q0 1.866-0.92 3.442t-2.496 2.496-3.442 0.92-3.442-0.92-2.496-2.496-0.92-3.442 0.92-3.442 2.496-2.496 3.442-0.92 3.442 0.92 2.496 2.496 0.92 3.442z"
          style="fill:#a4030f"
        />
      </svg>
    </div>
    <div
      class="alert-content emotion-1"
    >
      This is an error!
    </div>
  </div>
</div>
`;

exports[`verification of all variants of alert 1`] = `
Object {
  "backGroundColor": "#f2dede",
  "borderColor": "#ebccd1",
  "color": "#862422",
  "icon": <AlertErrorIcon
    fill="#a4030f"
  />,
}
`;

exports[`verification of all variants of alert 2`] = `
Object {
  "backGroundColor": "#fcf8e3",
  "borderColor": "#faebcc",
  "color": "#6f4f17",
  "icon": <AlertWarnIcon
    fill="#db781a"
  />,
}
`;

exports[`verification of all variants of alert 3`] = `
Object {
  "backGroundColor": "#dff0d8",
  "borderColor": "#d6e9c6",
  "color": "#215821",
  "icon": <AlertSuccessIcon
    fill="#6d9867"
  />,
}
`;

exports[`verification of all variants of alert 4`] = `
Object {
  "backGroundColor": "#d9edf7",
  "borderColor": "#b1dff3",
  "color": "#0e516f",
  "icon": <InfoIcon
    fill="#0271b9"
  />,
}
`;

exports[`verification of all variants of alert 5`] = `
Object {
  "backGroundColor": "#d9edf7",
  "borderColor": "#b1dff3",
  "color": "#0e516f",
  "icon": <DeferredSpinner
    timeout={0}
  />,
}
`;
